chainer/variable.py

```diff
@@ -511,6 +511,10 @@ class Variable(object):
     # TODO(hvy): Avoid modifying this variable from outside this class.
     _chainerx_fallback_array = None
 
+    _data = None  # type: tp.List[tp.Optional[types.NdArray]]
```

Comment:
Please remove type hints from internal interfaces and code.

Reply:
Hmm, it is required to remove type errors in the user-facing methods. How do you think about this kind of cases?

Reply:
Could you elaborate that? I thought internal attributes should not affect users' typing checks.

Reply:
If we specify types for arguments, mypy try to check if all variables (including the arguments) in the function scope are sound. It doesn't affect the user's type checking, but leaving these errors are not convenient because the type error output is useful for adding new type hints even if CI doesn't run type checking.

Reply:
Mmm, I don't still understand. What do you mean by it doesn't affect user's type checking? It sounds like it won't cause any errors.
Could you post actual commands?

Reply:
I tend to run `mypy chainer` in the `chainer` directory when adding type hints to Chainer for inspecting *type bugs*.

Reply:
Is that for checking chainer's internal type consistency?
As discussed previously (#6146), we're not going to guarantee such check would succeed.

Reply:
Yes, but sometimes, in my experience, the types of interface method are not obvious from the documentation. For example, I noticed #4074 independently because of the mypy error. In other case, the specification of `Initializer` [allows a scalar value](https://docs.chainer.org/en/stable/reference/initializers.html) but it is not explicitly documented in [some APIs](https://github.com/chainer/chainer/blob/7e4f072c6d42cd345ac216cfdf353bce891140e0/chainer/links/connection/linear.py#L34-L38). I noticed `InitializerSpec` should contain `ScalarValue` when running mypy.

I agree that non-public interfaces and variables should prevent from using type hints, but minimum usage is very useful for detecting problems described like above. How do you think?

Reply:
So do you mean you want to restart using mypy?
It's an option of course, but the team should agree with that first, as the current concensus is the opposite.

By the way, `L.Linear` docs has a link to initializers, so I'd say it's documented.

Reply:
Ok, I'm honestly wondering what means "use type hints only in the public interfaces," but I understood that the type hints basically should not be used in Chainer. I'll remove this hint and create a PR for the documentations. Thanks for the discussion.